### ISSUE TITLE
htop won't build on Ubuntu 14.04

### ISSUE THREAD

**Issue by hartzell** [spack/spack#8534]
**TL;DR**: htop seems to be expecting us to be installing our ncurses@6 libraries with a '6' suffix but we're not.  What's the proper fix?

## Summary

On a minimally configured (git, build-essentials, unzip) AWS Ubuntu 14.04 instance, Spack can not build *htop*.

```
hartzell@ubuntu-spack-test:~/spack$ spack install htop
==> pkgconf is already installed in /home/hartzell/spack/opt/spack/linux-ubuntu14.04-x86_64/gcc-5.5.0/pkgconf-1.4.2-w4jy3vpkv7mj4xx4swylm5ofrztzit3w
==> ncurses is already installed in /home/hartzell/spack/opt/spack/linux-ubuntu14.04-x86_64/gcc-5.5.0/ncurses-6.1-o6v4a6q52safi4w4x6mh7z2rwkzyjblr
==> Installing htop
==> Using cached archive: /home/hartzell/spack/var/spack/cache/htop/htop-2.0.2.tar.gz
==> Staging archive: /home/hartzell/spack/var/spack/stage/htop-2.0.2-n32omwlgkkuhp32h6gbt5rjw2bohmret/htop-2.0.2.tar.gz
==> Created stage in /home/hartzell/spack/var/spack/stage/htop-2.0.2-n32omwlgkkuhp32h6gbt5rjw2bohmret
==> No patches needed for htop
==> Building htop [AutotoolsPackage]
==> Executing phase: 'autoreconf'
==> Executing phase: 'configure'
==> Error: ProcessError: Command exited with status 1:
    '/home/hartzell/spack/var/spack/stage/htop-2.0.2-n32omwlgkkuhp32h6gbt5rjw2bohmret/htop-2.0.2/configure' '--prefix=/home/hartzell/spack/opt/spack/linux-ubuntu14.04-x86_64/gcc-5.5.0/htop-2.0.2-n32omwlgkkuhp32h6gbt5rjw2bohmret' '--enable-shared'

1 error found in build log:
     116    checking for addnwstr in -lncursesw6... no
     117    checking for addnwstr in -lncursesw... no
     118    checking for addnwstr in -lncurses... no
     119    checking for addnwstr in -lncursesw6... (cached) no
     120    checking for addnwstr in -lncursesw... (cached) no
     121    checking for addnwstr in -lncurses... (cached) no
  >> 122    configure: error: You may want to use --disable-unicode or install libncursesw.

See build log for details:
  /home/hartzell/spack/var/spack/stage/htop-2.0.2-n32omwlgkkuhp32h6gbt5rjw2bohmret/htop-2.0.2/spack-build.out
```

I see something similar was reported in #3126. 

*htop* does depend on ncurses:

```
hartzell@ubuntu-spack-test:~/spack$ spack spec htop
Input spec
--------------------------------
htop

Concretized
--------------------------------
htop@2.0.2%gcc@5.5.0 arch=linux-ubuntu14.04-x86_64
    ^ncurses@6.1%gcc@5.5.0~symlinks~termlib arch=linux-ubuntu14.04-x86_64
        ^pkgconf@1.4.2%gcc@5.5.0 arch=linux-ubuntu14.04-x86_64
```

and (unlike in #3126) the library with the `w` suffix is present:

```
hartzell@ubuntu-spack-test:~/spack$ find opt/ | grep libncursesw
opt/spack/linux-ubuntu14.04-x86_64/gcc-4.8/ncurses-6.1-mges4os4itr27x33pvonrcst22t5eroe/lib/libncursesw_g.a
opt/spack/linux-ubuntu14.04-x86_64/gcc-4.8/ncurses-6.1-mges4os4itr27x33pvonrcst22t5eroe/lib/libncursesw.a
opt/spack/linux-ubuntu14.04-x86_64/gcc-4.8/ncurses-6.1-mges4os4itr27x33pvonrcst22t5eroe/lib/libncursesw.so.6.1
opt/spack/linux-ubuntu14.04-x86_64/gcc-4.8/ncurses-6.1-mges4os4itr27x33pvonrcst22t5eroe/lib/libncursesw.so
opt/spack/linux-ubuntu14.04-x86_64/gcc-4.8/ncurses-6.1-mges4os4itr27x33pvonrcst22t5eroe/lib/libncursesw.so.6
opt/spack/linux-ubuntu14.04-x86_64/gcc-5.5.0/ncurses-6.1-o6v4a6q52safi4w4x6mh7z2rwkzyjblr/lib/libncursesw_g.a
opt/spack/linux-ubuntu14.04-x86_64/gcc-5.5.0/ncurses-6.1-o6v4a6q52safi4w4x6mh7z2rwkzyjblr/lib/libncursesw.a
opt/spack/linux-ubuntu14.04-x86_64/gcc-5.5.0/ncurses-6.1-o6v4a6q52safi4w4x6mh7z2rwkzyjblr/lib/libncursesw.so.6.1
opt/spack/linux-ubuntu14.04-x86_64/gcc-5.5.0/ncurses-6.1-o6v4a6q52safi4w4x6mh7z2rwkzyjblr/lib/libncursesw.so
opt/spack/linux-ubuntu14.04-x86_64/gcc-5.5.0/ncurses-6.1-o6v4a6q52safi4w4x6mh7z2rwkzyjblr/lib/libncursesw.so.6
hartzell@ubuntu-spack-test:~/spack$
```

Here's the test that failed from the `config.log`:

```
configure:12673: checking for addnwstr in -lncursesw6
configure:12698: /home/hartzell/spack/lib/spack/env/gcc/gcc -o conftest -g -O2  -L/home/hartzell/spack/opt/spack/linux-ubuntu14.04-x86_64/gcc-5.5.0/ncurses-6.1-o6v4a6q52safi4w4x6mh7z2rwkzyjblr/lib -lncursesw  conftest.c -lncursesw6  -lm  >&5
/usr/bin/ld: cannot find -lncursesw6
collect2: error: ld returned 1 exit status
```

Here's the bit of the configure script.  

```
if test "x$enable_unicode" = xyes; then

   if test ! -z "$HTOP_NCURSESW6_CONFIG_SCRIPT"; then
      # to be used to set the path to ncurses*-config when cross-compiling
      htop_config_script=$($HTOP_NCURSESW6_CONFIG_SCRIPT --libs 2> /dev/null)
   else
      htop_config_script=$("ncursesw6-config" --libs 2> /dev/null)
   fi
   htop_script_success=no
   htop_save_LDFLAGS="$LDFLAGS"
   if test ! "x$htop_config_script" = x; then
      LDFLAGS="$htop_config_script $LDFLAGS"
      { $as_echo "$as_me:${as_lineno-$LINENO}: checking for addnwstr in -lncursesw6" >&5
$as_echo_n "checking for addnwstr in -lncursesw6... " >&6; }
if ${ac_cv_lib_ncursesw6_addnwstr+:} false; then :
  $as_echo_n "(cached) " >&6
else
  ac_check_lib_save_LIBS=$LIBS
LIBS="-lncursesw6  $LIBS"
cat confdefs.h - <<_ACEOF >conftest.$ac_ext
/* end confdefs.h.  */
```

It's finding and calling `ncursesw6-config` successfully (which give it *both* the `-L` and `-l` that it's using, but then it goes on and tacks on it's own `-lncursesw6`.  We don't install a copy of the library with the `6` ending.

Setting `force_autoreconf=True` does not work.

This edit of the configure script *does* work:

```
--- configure.ac.orig   2018-06-20 22:10:44.012015788 +0000
+++ configure.ac        2018-06-20 22:10:57.304015788 +0000
@@ -185,7 +185,7 @@

 AC_ARG_ENABLE(unicode, [AS_HELP_STRING([--enable-unicode], [enable Unicode support])], ,enable_unicode="yes")
 if test "x$enable_unicode" = xyes; then
-   HTOP_CHECK_SCRIPT([ncursesw6], [addnwstr], [HAVE_LIBNCURSESW], "ncursesw6-config",
+   HTOP_CHECK_SCRIPT([ncursesw], [addnwstr], [HAVE_LIBNCURSESW], "ncursesw6-config",
     HTOP_CHECK_SCRIPT([ncursesw], [addnwstr], [HAVE_LIBNCURSESW], "ncursesw5-config",
      HTOP_CHECK_SCRIPT([ncurses], [addnwstr], [HAVE_LIBNCURSESW], "ncurses5-config",
       HTOP_CHECK_LIB([ncursesw6], [addnwstr], [HAVE_LIBNCURSESW],
```

Is the fix to patch htop or to install ncursesw libs with a '6' suffix?

**Comment by adamjstewart** [comment]
> Is the fix to patch htop or to install ncursesw libs with a '6' suffix?

I would err on the side of patching `htop`. Otherwise we'll end up in the hell of building all possible combinations of `libncurses`, `libncursesw`, `libncurses6`, and `libncursesw6`.

**Comment by hartzell** [comment]
The project that was looking at becoming Ubuntu based is going to remain on CentOS, so this isn't relevant (to me) for now.

I can close this or leave it open, but I don't see myself running with it.

Suggestions?

**Comment by adamjstewart** [comment]
In light of #8811 lets just close it. It's still searchable if someone else wants to reopen it.
